### PR TITLE
MAINT: Switch np.test to pytest

### PR DESCRIPTION
**Pull request by ghost**
This is getting more into the area of compatibility breakage, so it needs discussion about what existing expectations are.

### CODE REVIEW

**Comment by charris** [comment]
Are any of the tests still using nose? I figure that at some point we need to make a clean break for NumPy testing, the compatibility was just to make it easier taking our time getting there. For backwards compatibility we just need to worry about downstream projects that may be using the NumPy testing framework. When we do make the break `runtests.py` needs to be working with pytest, and that by itself will impact compatibility and also the tests done during wheels generation.

**Comment by charris** [comment]
It would be could if you could be more careful with commit messages. Maybe clean up the history when this is ready.

**Comment by xoviat** [comment]
Let me be clear: merging this will drop support for testing NumPy with nose (though the nose tooling will still be present). The *only* question is whether there is unacceptable breakage with the `np.test` and `np.bench` functions.

**Comment by charris** [comment]
Well, they should both still work, along with `python runtests.py ...`. What do you mean by breakage?

**Comment by njsmith** [comment]
> (though the nose tooling will still be present)

So this *won't* break other projects that are using `np.testing.NoseTester`?

**Comment by xoviat** [comment]
If they're using it to test their own project with nose, then no.

**Comment by xoviat** [comment]
This PR needs more attention. Which downstream projects are using nosetester?

**Comment by pv** [comment]
I think the main approach in this PR is fine, i.e., just switch Numpy's 
tests to run on pytest, and leave numpy.testing alone.

Did you check whether the number of tests run remains the same as before 
this PR, i.e., do all the tests run?

On backward compat: since this PR doesn't seem to touch any public parts 
of numpy.testing (which come from numpy.testing.nose_tools), the only 
backward compat question is indeed numpy.test and numpy.bench. I would 
imagine the main automated users of `numpy.test` are various 
distributions and packagers (Debian, Anaconda, et al.), and they are 
able to deal with such unannounced version-dependent changes. Other 
users probably just run the test suite manually, so that minor behavior 
differences don't matter.

With scipy we just broke exact `scipy.test` backward compatibility cold 
turkey and didn't receive much complaints.

I see the `numpy/lib/_testutils.py` is copypasted from Scipy. It's maybe 
useful to remove the unused helper functions.

The `numpy.bench` function doesn't do anything even now, and could 
probably even be removed.

**Comment by xoviat** [comment]
@pv How did SciPy keep the doctests running. Does NumPy have doctests and/or are they currently run?

**Comment by pv** [comment]
@xoviat: scipy has an ugly thing that runs them: https://github.com/scipy/scipy/blob/master/tools/refguide_check.py

File: tools/ci/appveyor/requirements.txt [hunk 1]
```diff
@@ -4,3 +4,4 @@ pytest-timeout
 pytest-xdist
 pytest-env
 pytest-faulthandler
+pytest-warnings
```
Comment (pv) [inline tools/ci/appveyor/requirements.txt]: iiuc this is builtin in pytest >= 3.1

**Comment by xoviat** [comment]
I think this is ready to merge?

cc @eric-wieser

**Comment by charris** [comment]
@ghost Could you add a mention to the 1.15.0 release notes? Also, I assume we need some small changes to `numpy-wheels`, if only installing pytest. I'm also going to be adding a Python 3.7, Cython 0.28.0 test soon, the new Cython release should be out sometime today or tomorrow.

**Comment by pv** [comment]
@charris: "ghost" is a placeholder for a deleted account (xoviat, who as I read elsewhere on github, is taking a break from github stuff).

**Comment by charris** [comment]
I knew it was "xoviat". Hmm, how did he push commits 13 days ago? Maybe he is just blocked.

**Comment by charris** [comment]
Nose is going to be a problem with Python 3.7 and stop working in 3.8, so we need to finish this up.

**Comment by charris** [comment]
So this cannot be downloaded from github, and also brings back `unittest` which I don't like. Hmm...

**Comment by pv** [comment]
@charris: you can get it with

git fetch upstream refs/pull/10379/head:pr-10379

You may also want to add the line

fetch = +refs/pull/*/head:refs/remotes/upstream/pr/*

to your git config [remote "upstream"] section so that "git fetch"
brings all pull requests to you locally as branches named
"upstream/pr/XXX".

**Comment by pv** [comment]
cf.
https://help.github.com/articles/checking-out-pull-requests-locally/
https://gist.github.com/piscisaureus/3342247

**Comment by charris** [comment]
Closing in favor of #10768.

**Comment by charris** [comment]
Thanks @pv, I've put it up as #10768.

**Pull request closed without merging**